### ISSUE TITLE
[clang static analyzer] core.NullDereference false positive with `*p = 42`

### ISSUE THREAD

**Issue by 0-0x41** [llvm/llvm-project#60025]
I got a false positive error when compiling the following program with clang(trunk) `--analyze -Xclang -analyzer-stats -Xclang -analyzer-checker=core,debug.ExprInspection` in https://godbolt.org/z/nEMcfT883.

In this case, the judgment condition of the if statement should be FALSE and its inner code is unreachable code. However, in line 11, analyzer gives FALSE for `(d[0].b && 1) == false`. This is a little odd. Here is the analysis result of the case. Thank you for taking the time to review this case.

Input:

```c
#include "stdio.h"
#include <stdbool.h>
void clang_analyzer_eval();

union a
{
    int b
} c()
{
    union a d[1] = {0};
    int *p = (int *)0;
    clang_analyzer_eval((d[0].b && 1) == false);
    if (d[0].b && 1)
    {
        *p = 42;
    }
}

```

Output:

```bash
<source>:11:5: warning: FALSE [debug.ExprInspection]
    clang_analyzer_eval((d[0].b && 1) == false);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:14:12: warning: Dereference of null pointer (loaded from variable 'p') [core.NullDereference]
        *p = 42;

```
